File: Python3/1 - 200/79 Word Search.py
# 316ms, 86.28%; 13.2MB, 100.00%
from typing import List, Tuple


class Solution:  # quite similar with #797
    def exist(self, board: List[List[str]], word: str) -> bool:
        start = self.__find(board, word[0])
        if not start: return False  # first char not exists or board is empty
        row, col = len(board), len(board[0])
        st, visited = [start], [[False] * col for _ in range(row)]
        while st:
            if len(st) == len(word): return True  # completed the word
            i, j = st[-1][-1]  # add this one into 'stack'
            visited[i][j] = True
            next_step = self.__find(board, word[len(st)], st[-1][-1], visited)
            if next_step:  # (i, j) is correct so far
                st.append(next_step)
            else:  # cannot continue, then rollback the stack and visit status
                while st and len(st[-1]) == 1:
                    x, y = st.pop()[0]
                    visited[x][y] = False
                if st:
                    x, y = st[-1].pop()
                    visited[x][y] = False
        return False

    def __find(self,
               board: List[List[str]],
               char: str,
               hint: Tuple[int] = None,
               visited: List[List[bool]] = None) -> List[Tuple[int]]:
        ret = []
        if hint:  # check at most four adjacent cells
            i, j = hint
            if i and board[i - 1][j] == char and not visited[i - 1][j]: ret.append((i - 1, j))
            if j and board[i][j - 1] == char and not visited[i][j - 1]: ret.append((i, j - 1))
            if i + 1 < len(board) and board[i + 1][j] == char and not visited[i + 1][j]: ret.append((i + 1, j))
            if j + 1 < len(board[0]) and board[i][j + 1] == char and not visited[i][j + 1]: ret.append((i, j + 1))
        else:  # find start points
            for i in range(len(board)):
                for j in range(len(board[0])):
                    if board[i][j] == char: ret.append((i, j))
        return ret

File: 1 - 200/79 Word Search.cc
// 68ms, 52.50%; 16.6MB, 51.47%
#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    typedef vector<pair<int, int>> vpair;
    vector<vector<bool>> isVisited;
    pair<int, int> hint;

public:
    bool exist(const vector<vector<char>> &board, const string &word) {
        vpair start = find(board, word[0]);
        if (start.empty()) return false;
        int row = board.size(), col = board[0].size(), length = word.size();
        isVisited = vector<vector<bool>>(row, vector<bool>(col, false));
        stack<vpair> st;
        st.push(start);
        while (!st.empty()) {
            if (st.size() == length) return true;
            hint = st.top().back();
            isVisited[hint.first][hint.second] = true;
            vpair next_step = find(board, word[st.size()], true);
            if (!next_step.empty()) {
                st.push(move(next_step));
            } else {
                while (!st.empty() && st.top().size() == 1) {
                    int x = st.top()[0].first, y = st.top()[0].second;
                    st.pop(), isVisited[x][y] = false;
                }
                if (!st.empty()) {
                    int x = st.top().back().first, y = st.top().back().second;
                    st.top().pop_back(), isVisited[x][y] = false;
                }
            }
        }
        return false;
    }

private:
    vpair find(const vector<vector<char>> &board, char ch, bool useHint = false) {
        vpair ret;
        if (useHint) {
            int i = hint.first, j = hint.second;
            if (i && board[i - 1][j] == ch && !isVisited[i - 1][j]) ret.emplace_back(i - 1, j);
            if (j && board[i][j - 1] == ch && !isVisited[i][j - 1]) ret.emplace_back(i, j - 1);
            if (i + 1 < board.size() && board[i + 1][j] == ch && !isVisited[i + 1][j]) ret.emplace_back(i + 1, j);
            if (j + 1 < board[0].size() && board[i][j + 1] == ch && !isVisited[i][j + 1]) ret.emplace_back(i, j + 1);
        } else {
            for (int i = 0, row = board.size(); i < row; ++i)
                for (int j = 0, col = board[0].size(); j < col; ++j)
                    if (board[i][j] == ch) ret.emplace_back(i, j);
        }
        return ret;
    }
};
